gobblin-core/src/main/java/gobblin/writer/http/AbstractAsyncDataWriter.java

```diff
@@ -0,0 +1,52 @@
+package gobblin.writer.http;
+
+import java.io.IOException;
+import java.util.concurrent.Future;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import gobblin.async.AsyncDataDispatcher;
+import gobblin.writer.AsyncDataWriter;
+import gobblin.writer.WriteCallback;
+import gobblin.writer.WriteResponse;
+
+
+/**
+ * Base class to write data asynchronously. It is an {@link AsyncDataDispatcher} on {@link BufferedRecord}, which
+ * wraps a record and its callback.
+ *
+ * @param <D> type of record
+ */
+@ThreadSafe
+public abstract class AbstractAsyncDataWriter<D> extends AsyncDataDispatcher<BufferedRecord<D>> implements AsyncDataWriter<D> {
+  public static final int DEFAULT_BUFFER_CAPACITY = 10000;
+
+  public AbstractAsyncDataWriter(int capacity) {
+    super(capacity);
+  }
+
+  @Override
+  public Future<WriteResponse> write(D record, @Nullable WriteCallback callback) {
+    BufferedRecord<D> bufferedRecord = new BufferedRecord<>(record, callback);
+    put(bufferedRecord);
+    return null;
```

Comment:
We'll use the callback. It's not necessary to maintain 2 kinds of async semantics at the same time. I'll add some comments.